tests/psbt.rs

```diff
@@ -0,0 +1,704 @@
+    let expected_psbt_hex = include_str!("data/sign_2_psbt_hex");
+    let expected_psbt = hex_psbt!(expected_psbt_hex).unwrap();
+
+    let psbt = sign(psbt, keys);
+
+    assert_eq!(psbt, expected_psbt);
+    psbt
+}
+
+/// Does the combine according to the BIP, returns the combined PSBT. Verifies against BIP174 test vector.
+fn combine(mut this: Psbt, that: Psbt) -> Psbt {
+    let expected_psbt_hex = include_str!("data/combine_psbt_hex");
+    let expected_psbt = hex_psbt!(expected_psbt_hex).unwrap();
+
+    this.combine(that).expect("failed to combine PSBTs");
+
+    assert_eq!(this, expected_psbt);
+    this
+}
+
+/// Does the finalize step according to the BIP, returns the combined PSBT. Verifies against BIP174
+/// test vector.
+fn finalize(psbt: Psbt) -> Psbt {
+    let expected_psbt_hex = include_str!("data/finalize_psbt_hex");
+    let expected_psbt = hex_psbt!(expected_psbt_hex).unwrap();
+
+    let psbt = finalize_psbt(psbt);
+
+    assert_eq!(psbt, expected_psbt);
+    psbt
+}
+
+/// Does the transaction extractor step according to the BIP, returns the combined PSBT. Verifies
+/// against BIP174 test vector.
+fn extract_transaction(psbt: Psbt) -> Transaction {
+    let expected_tx_hex = include_str!("data/extract_tx_hex");
+
+    let tx = psbt.extract_tx();
+    
+    let got = serialize_hex(&tx);
+    assert_eq!(got, expected_tx_hex);
+
+    tx
+}
+
+/// Combines two PSBTs lexicographically according to the BIP. Verifies against BIP174 test vector.
+fn combine_lexicographically() {
+    let psbt_1_hex = include_str!("data/lex_psbt_1_hex");
+    let psbt_2_hex = include_str!("data/lex_psbt_2_hex");
+
+    let expected_psbt_hex = include_str!("data/lex_combine_psbt_hex");
+    let expected_psbt = hex_psbt!(expected_psbt_hex).unwrap();
+    
+    let v = Vec::from_hex(&psbt_1_hex).unwrap();
+    let mut psbt_1: Psbt = deserialize(&v).expect("failed to deserialize psbt 1");
+
+    let v = Vec::from_hex(&psbt_2_hex).unwrap();
+    let psbt_2: Psbt = deserialize(&v).expect("failed to deserialize psbt 2");
+
+    psbt_1.combine(psbt_2).expect("failed to combine PSBTs");
+    
+    assert_eq!(psbt_1, expected_psbt);
+}
+
+/// Signs `psbt` with `keys` if required.
+fn sign(mut psbt: Psbt, keys: &[PrivateKey]) -> Psbt {
+    let secp = SECP256K1;
+
+    let mut inputs = psbt.inputs.clone();
+
+    for (i, input) in inputs.iter_mut().enumerate() {
+        let required_pks: Vec<_> = input.bip32_derivation.keys().collect();
+        for sk in keys {
+            let pk = sk.public_key(&secp);
+            if required_pks.contains(&&pk.inner) {
+                sign_psbt(&sk, &mut psbt, i).expect("failed to sign input with key");
+            }
+        }
+    }
+
+    psbt
+}
+
+/// Signs `psbt` input at `input_index` using `sk`.
+// TODO: Consider adding this code to rust-bitcoin.
+fn sign_psbt(
+    sk: &PrivateKey,
+    psbt: &mut Psbt,
+    input_index: usize,
+) -> Result<(), SignError> {
```

Comment:
Typically tests don't have errors in them. This code is handy, we should use it in the main repo. This would need to be updated for taproot tough.

Reply:
Agree, this error exists because of the as-yet-unresolved desire to upstream the bdk code. I'm starting to feel like this PR and #940 should be draft until #946 resolves?

Reply:
#946 closed replaced by #957

Reply:
> This would need to be updated for taproot tough.

After spending today on taproot signing, I got a giggle out of this typo :)